libraries/AP_HAL_Linux/CAN.cpp

```diff
@@ -0,0 +1,612 @@
+            return 0;
+        }
+        return res;
+    }
+    if (res != sizeof(sockcan_frame)) {
+        return -1;
+    }
+    return 1;
+}
+
+
+int LinuxCAN::_read(uavcan::CanFrame& frame, uavcan::UtcTime& ts_utc, bool& loopback) const
+{
+    auto iov = iovec();
+    auto sockcan_frame = can_frame();
+    iov.iov_base = &sockcan_frame;
+    iov.iov_len  = sizeof(sockcan_frame);
+
+    struct Control
+    {
+        cmsghdr cm;
+        std::uint8_t data[sizeof(::timeval)];
+    };
+    auto control = Control();
+
+    auto msg = msghdr();
+    msg.msg_iov    = &iov;
+    msg.msg_iovlen = 1;
+    msg.msg_control = &control;
+    msg.msg_controllen = sizeof(control);
+
+    const int res = recvmsg(_fd, &msg, MSG_DONTWAIT);
+    if (res <= 0) {
+        return (res < 0 && errno == EWOULDBLOCK) ? 0 : res;
+    }
+    /*
+     * Flags
+     */
+    loopback = (msg.msg_flags & static_cast<int>(MSG_CONFIRM)) != 0;
+
+    if (!loopback && !_checkHWFilters(sockcan_frame)) {
+        return 0;
+    }
+
+    frame = makeUavcanFrame(sockcan_frame);
+    /*
+     * Timestamp
+     */
+    const cmsghdr* const cmsg = CMSG_FIRSTHDR(&msg);
+    if (cmsg->cmsg_level == SOL_SOCKET && cmsg->cmsg_type == SO_TIMESTAMP) {
+        auto tv = timeval();
+        std::memcpy(&tv, CMSG_DATA(cmsg), sizeof(tv));  // Copy to avoid alignment problems
+        ts_utc = uavcan::UtcTime::fromUSec(std::uint64_t(tv.tv_sec) * 1000000ULL + tv.tv_usec);
+    } else {
+        return -1;
+    }
+    return 1;
+}
+
+void LinuxCAN::_incrementNumFramesInSocketTxQueue()
+{
+    _frames_in_socket_tx_queue++;
+}
+
+void LinuxCAN::_confirmSentFrame()
+{
+    if (_frames_in_socket_tx_queue > 0) {
+        _frames_in_socket_tx_queue--;
+    }
+}
+
+bool LinuxCAN::_wasInPendingLoopbackSet(const uavcan::CanFrame& frame)
+{
+    if (_pending_loopback_ids.count(frame.id) > 0) {
+        _pending_loopback_ids.erase(frame.id);
+        return true;
+    }
+    return false;
+}
+
+bool LinuxCAN::_checkHWFilters(const can_frame& frame) const
+{
+    if (!_hw_filters_container.empty()) {
+        for (auto& f : _hw_filters_container) {
+            if (((frame.can_id & f.can_mask) ^ f.can_id) == 0) {
+                return true;
+            }
+        }
+        return false;
+    } else {
+        return true;
+    }
+}
+
+void LinuxCANDriver::IfaceWrapper::updateDownStatusFromPollResult(const pollfd& pfd)
+{
+    if (!down_ && (pfd.revents & POLLERR)) {
+        int error = 0;
+        socklen_t errlen = sizeof(error);
+        getsockopt(pfd.fd, SOL_SOCKET, SO_ERROR, reinterpret_cast<void*>(&error), &errlen);
+
+        down_ = error == ENETDOWN || error == ENODEV;
+
+        hal.console->printf("Iface %d is dead; error %d", this->getFileDescriptor(), error);
+    }
+}
+
+void LinuxCANDriver::register_thread() {
+    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&LinuxCANDriver::_timer_tick, void));
```

Comment:
I think we should have a real thread instead of hooking to the timer (see the Thread abstraction in the Linux HAL... PollerThread would be ideal for this since it has a full featured epoll abstraction).

The architecture here is a bit odd in which you register a callback here, then you call out to `p_uavcan->do_cyclic()`, which calls back into here... but it's doable with PollerThread, too. Long term I'd like to see users of timer to go away.